Refactor usage of implicit parameter to an explicit one when rendering

ibesora opened the pull request:
This PR aims to refactor our codebase a little bit in order to move away of what, IMHO, are code smells.

First, some context. 
The [Painter](https://github.com/maplibre/maplibre-gl-js/blob/main/src/render/painter.ts#L69) object mostly contains properties that describe how things are going to be rendered via the underlying rendering API (i.e. vertex and index buffers, stencil buffer definitions, depth range, etc.) but it also contains the `Style` object which in itself describes how things should look like. 
`draw_*` functions (i.e. [draw_background](https://github.com/maplibre/maplibre-gl-js/blob/b6e3b551e85b4fe42211226f75c0991cba698418/src/render/draw_background.ts#L15)) get both the `Painter` object and `StyleLayer` object, which is basically the subset of the `Style` that we need to render a given layer.

There are various places in the `draw_*` functions where we use `painter.style` to get information about the current rendering mode:
- We use `painter.style.map.terrain` to know if we are rendering with terrain
- We use `painter.style.projection.name` to know if we are rendering using `globe` or `mercator` projections

and change the drawing behaviour according to that. Since those properties are deep into the `Painter` object, you might get two similar-looking `draw_*` calls with very different results. 

This PR replaces checks for terrain rendering via the `painter.style` property with an extra `renderLayer` parameter that tells a given `draw_*` function if it's being rendered to a texture.

A follow-up idea would be to remove `style` entirely from the `Painter` object and add all the properties that are needed for the various `draw_*` functions into the `Painter` object itself.

Cc @birkskyum since we talked about this.

## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 - [x] Briefly describe the changes in this PR.
 - [ ] Link to related issues.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
 - [ ] Write tests for all new functionality.
 - [ ] Document any changes to public APIs.
 - [ ] Post benchmark scores.
 - [ ] Add an entry to `CHANGELOG.md` under the `## main` section.

birkskyum commented:
@ibesora , thanks! I think this is much better. I have an ask - is there a way we can flip this around? So instead of enabling lots of matrix transformations by defualt, and then actively disabling the ones we don't want, that we instead pass the transformations that we **_do_** want?

HarelM commented:
I think this is great!
I would also look in the code and find the places where "this" is being passed down. Solving those places will prevent this behavior I believe where you look deep into objects to find references to other objects. 

Since we are in the middle of a breaking change version release, now would be a great time to move objects around and clean some historical bad decisions...

HarelM commented on src/render/painter.ts:
```diff
@@ -640,34 +640,34 @@ export class Painter {
                 break;
             case 'custom':
-                drawCustom(painter, sourceCache, layer as any);
+                drawCustom(painter, sourceCache, layer as any, isRenderingToTexture);
```
Can you fix all these "as any" while you change this code?

birkskyum commented on src/render/draw_heatmap.ts:
```diff
@@ -179,7 +179,7 @@ function renderHeatmapTerrain(painter: Painter, layer: HeatmapStyleLayer, coord:
     colorRampTexture.bind(gl.LINEAR, gl.CLAMP_TO_EDGE);
 
-    const projectionData = transform.getProjectionData({overscaledTileID: coord, ignoreTerrainMatrix: !isGlobe});
+    const projectionData = transform.getProjectionData({overscaledTileID: coord, ignoreTerrainMatrix: !isGlobe, ignoreGlobeMatrix: isRenderingToTexture});
```
Can we do the same for ignoreTerrainMatrix ? 


```ts
const isGlobe = painter.style.projection.name === 'globe';

ignoreTerrainMatrix: !isGlobe,
```

birkskyum replied:
We need to pull out the check on the "globe" string for anywhere that uses the globe, which is all ths, because the projection "vertical-perspective" is also using the same logic, just without the transition to mercator.

birkskyum commented:
We need to pull out the check on the "globe" string for anywhere that uses the globe, because the projection "vertical-perspective" is also using the same logic, just without the transition to mercator.

ibesora commented:
I've updated this PR with some of your suggestions:
- To remove reading `globe` from the style as suggested [here](https://github.com/maplibre/maplibre-gl-js/pull/5020#discussion_r1835143304), instead of adding another boolean to all the `draw_*` functions, I moved those into a type called `RenderFlags` that will be able to grow with more flags.
- I also added type predicate functions to change remove the `layer as any` casts.

I'll tackle removing `style` from the `Painter` object in another PR

HarelM commented on src/render/painter.ts:
```diff
@@ -62,6 +72,11 @@ type PainterOptions = {
     fadeDuration: number;
 };
 
+export type RenderFlags = {
```
I think "Options" is used in the code. I would consider renaming this to "RenderOptions".

HarelM commented on src/render/painter.ts:
```diff
@@ -633,41 +649,41 @@ export class Painter {
-        switch (layer.type) {
-            case 'symbol':
-                drawSymbols(painter, sourceCache, layer as any, coords, this.style.placement.variableOffsets);
+        switch (true) {
```
I find this odd TBH, maybe a if ...else if... Will be better here? It might be just that I'm not used to it... IDK.

ibesora replied:
Yeah, to be fair it was a little bit weird. Changed it to an `if...else` block and it feels better.

HarelM commented on src/render/draw_background.ts:
```diff
@@ -6,13 +6,13 @@ import {
 import {coveringTiles} from '../geo/projection/covering_tiles';
 
-export function drawBackground(painter: Painter, sourceCache: SourceCache, layer: BackgroundStyleLayer, coords: Array<OverscaledTileID>, renderFlags: RenderFlags) {
+export function drawBackground(painter: Painter, sourceCache: SourceCache, layer: BackgroundStyleLayer, coords: Array<OverscaledTileID>, renderFlags: RenderOptions) {
```
Don't forget to change the parameter name as well - renderOptions...

HarelM commented:
LGTM, thanks!

birkskyum commented on src/geo/projection/globe_transform.test.ts:
```diff
@@ -46,14 +46,14 @@ describe('GlobeTransform', () => {
             expectToBeCloseToArray(projectionData.tileMercatorCoords, [0.5, 0, 0.5 / EXTENT, 0.5 / EXTENT]);
         });
 
-        test('Globe transition is not 0 when not ignoring the globe matrix', () => {
```
Thanks you, this is a lot cleaner.

From Triple negation
`        test('Globe transition is not 0 when not ignoring the globe matrix', () => {`

To Single negation
`         test('Globe transition is 0 when not applying the globe matrix', () => {`

birkskyum commented on src/render/draw_background.ts:
```diff
@@ -47,7 +47,7 @@ export function drawBackground(painter: Painter, sourceCache: SourceCache, layer
         const projectionData = transform.getProjectionData({
             overscaledTileID: tileID,
-            ignoreGlobeMatrix: globeWithTerrain
+            applyGlobeMatrix: !isRenderingToTexture
```
is an `applyTerrainMaterix: true` missing here?

birkskyum replied:
Or maybe background isn't affected by terrain?

birkskyum commented on src/render/draw_circle.ts:
```diff
@@ -80,7 +81,7 @@ export function drawCircles(painter: Painter, sourceCache: SourceCache, layer: C
         const uniformValues = circleUniformValues(painter, tile, layer, translateForUniforms, radiusCorrectionFactor);
 
-        const projectionData = transform.getProjectionData({overscaledTileID: coord});
+        const projectionData = transform.getProjectionData({overscaledTileID: coord, applyGlobeMatrix: !isRenderingToTexture});
```
Is an applyTerrainMatrix missing here too?

birkskyum commented on src/render/draw_fill.ts:
```diff
@@ -106,10 +108,9 @@ function drawFillTiles(
         const projectionData = transform.getProjectionData({
             overscaledTileID: coord,
-            ignoreGlobeMatrix: globeWithTerrain
+            applyGlobeMatrix: !isRenderingToTexture
```
Here also, I'd expect applyTerrainMatrix: true

birkskyum commented on src/render/draw_hillshade.ts:
```diff
@@ -73,12 +75,10 @@ function renderHillshade(
             overscaledTileID: coord,
             aligned: align,
-            ignoreGlobeMatrix: globeWithTerrain
+            applyGlobeMatrix: !isRenderingToTexture
```
here too, applyTerrainMatrix: true
